### PR TITLE
Fixed the DOI connection Error by Automatically Removing all Special Characters at the end of DOI

### LINKED ISSUE #8127: Invalid DOI should not show "Connection error"
### JabRef version

Latest development branch build (please note build date below)

### Operating system

Windows

### Details on version and operating system

Windows 10

### Checked with the latest development build

- [X] I made a backup of my libraries before testing the latest development version.
- [X] I have tested the latest development version and the problem persists

### Steps to reproduce the behaviour

1. Open https://link.springer.com/book/10.1007/978-3-319-47590-5
2. Download PDF ("Download book PDF")
3. Go to page 207
4. Select and copy the DOI:  
  image
5. Open "New entry dialog" and paste DOI:  
  image
6. See the question mark at the end
7. Click "Generate"
8. Result:  
  image

Expected behavior:

- The invalid character at the end should be removed internally
- The DOI data should be fetched

### Appendix

_No response_

### PR DESCRIPTION
**Pull request by jiezheng5**
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by jiezheng5** [comment]
I fixed all the checkstyle errors.

File: src/main/java/org/jabref/model/entry/identifier/DOI.java [hunk 1]
```diff
@@ -141,6 +145,47 @@ public DOI(String doi) {
...
+                    lcharDoi[i] == 59 ||
+                    lcharDoi[i] == 40 ||
+                    lcharDoi[i] == 41
+            ) {
+                // System.out.println(i);
+                break;
+            }
+        }
+
+//        char[] lcharDoi = doiStr.toCharArray();
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/entry/identifier/DOI.java:174]: remove commented out code

Comment (mrcstan) [reply src/main/java/org/jabref/model/entry/identifier/DOI.java:174]: This is not really in the spirit of open source contribution. You requested to work on this issue after me and I have started working on it.

Comment (jiezheng5) [reply src/main/java/org/jabref/model/entry/identifier/DOI.java:174]: > This is not really in the spirit of open source contribution. You requested to work on this issue after me and I have started working on it.

I actually started working on this issue 1 week ago. Then I saw your request, I thought maybe I should also post my progress. Do we have to first get permission before we can work on the open issues? 

thanks,

Comment (jiezheng5) [reply src/main/java/org/jabref/model/entry/identifier/DOI.java:174]: > remove commented out code

thanks very much for the suggestions, will do.

Comment (koppor) [reply src/main/java/org/jabref/model/entry/identifier/DOI.java:174]: > and I have started working on it.

@mrcstan May we see your approach? Maybe it differs significantly of this one - and we can make use of your ideas, too?

**Review by koppor — CHANGES_REQUESTED** [review]

**Comment by koppor** [comment]
Fixes https://github.com/JabRef/jabref/issues/8127

**Comment by koppor** [comment]
With more than 3 core developers, we looked at this issue in today's dev call. We found out, that there is `DOI.findInText` which resolves the issue --> https://github.com/JabRef/jabref/pull/8227

@jiezheng5 Could you add your test cases to that PR? I would like to ask you to add a new test section below `// findDoiInsideArbitraryText` with your test cases of this branch.

**Comment by koppor** [comment]
I am closing this one, as we think, the tests should be merged with #8227

**Comment by mrcstan** [comment]
I guess I just missed this one. I created a PR that uses regexp [#8288](https://github.com/JabRef/jabref/pull/8228). Seems like it has been fixed elsewhere. Let me know if I can contribute anything else.

File: .idea/runConfigurations/JabRef_Main.xml [hunk 2]
```diff
@@ -2,7 +2,7 @@
...
...e.control=org.controlsfx.controls --add-exports javafx.controls/com.sun.javafx.scene.control.inputmap=org.controlsfx.controls --add-exports javafx.base/com.sun.javafx.event=org.controlsfx.controls --add-exports javafx.base/com.sun.javafx.collections=org.controlsfx.controls --add-exports javafx.base/com.sun.javafx.runtime=org.controlsfx.controls --add-exports javafx.web/com.sun.webkit=org.controlsfx.controls --add-exports javafx.graphics/com.sun.javafx.css=org.controlsfx.controls --add-opens javafx.controls/javafx.scene.control.skin=org.controlsfx.controls --add-opens javafx.graphics/javafx.scene=org.controlsfx.controls --add-exports com.oracle.truffle.regex/com.oracle.truffle.regex=org.graalvm.truffle --add-exports javafx.graphics/com.sun.javafx.stage=com.jfoenix --add-exports javafx.controls/com.sun.javafx.scene.control.behavior=com.jfoenix -ea --add-exports=javafx.controls/com.sun.javafx.scene.control=org.jabref --add-exports=org.controlsfx.controls/impl.org.controlsfx.skin=org.jabref --add-exports javafx.graphics/com.sun.javafx.application=ALL-UNNAMED --add-exports javafx.graphics/com.sun.javafx.stage=ALL-UNNAMED --add-exports javafx.graphics/com.sun.javafx.stage=com.jfoenix" />
```
[the remark this answers is not in the document]
Comment (jiezheng5) [reply .idea/runConfigurations/JabRef_Main.xml:5]: > May I ask of the reasons why this file mas modified? IntelliJ runs fine as is.

Hi, Kopper:

Thanks very much for the feedback. I think IntelliJ automatically did it. I had lots of trouble running using IntelliJ both on windows and Linux. So I just used gradlew to run, which worked perfectly.

Thanks

File: CHANGELOG.md [hunk 3]
```diff
@@ -44,6 +44,8 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
 
 ### Fixed
 
+- We fixed an issue where an invalid DOI should not show "Connection error" [#8127](https://github.com/JabRef/jabref/issues/8127)
+- We fixed an issue where an exception ocurred when a linked online file was edited in the entry editor [#8008](https://github.com/JabRef/jabref/issues/8008)
```
[the remark this answers is not in the document]
Comment (jiezheng5) [reply CHANGELOG.md:49]: sure, will do. Thanks very much.

File: src/main/java/org/jabref/model/entry/identifier/DOI.java [hunk 4]
```diff
@@ -106,6 +106,10 @@ public DOI(String doi) {
         // Remove whitespace
         String trimmedDoi = doi.trim();
 
+        // jie add-------------------------
```
[the remark this answers is not in the document]
Comment (jiezheng5) [reply src/main/java/org/jabref/model/entry/identifier/DOI.java]: thanks very much for your patience, I forgot to remove it after debugging. will do.

have a great night.

File: src/main/java/org/jabref/model/entry/identifier/DOI.java [hunk 5]
```diff
@@ -141,6 +145,47 @@ public DOI(String doi) {
...
+                    (lcharDoi[i] >= 97 && lcharDoi[i] <= 122) ||
+                    lcharDoi[i] == 95 ||
+                    lcharDoi[i] == 59 ||
+                    lcharDoi[i] == 40 ||
+                    lcharDoi[i] == 41
+            ) {
+                // System.out.println(i);
+                break;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (jiezheng5) [reply src/main/java/org/jabref/model/entry/identifier/DOI.java]: ok, great suggestion. I will refactor the code and test the RegEx.

greatly appreciated all the suggestions

**Comment by jiezheng5** [comment]
> With more than 3 core developers, we looked at this issue in today's dev call. We found out, that there is `DOI.findInText` which resolves the issue --> #8227
> 
> @jiezheng5 Could you add your test cases to that PR? I would like to ask you to add a new test section below `// findDoiInsideArbitraryText` with your test cases of this branch.

ok.

**Comment by jiezheng5** [comment]
> With more than 3 core developers, we looked at this issue in today's dev call. We found out, that there is `DOI.findInText` which resolves the issue --> #8227
> 
> @jiezheng5 Could you add your test cases to that PR? I would like to ask you to add a new test section below `// findDoiInsideArbitraryText` with your test cases of this branch.

my pleasure. will do.

**Pull request closed without merging**